### PR TITLE
[DNM] Implements IP utilization tracking for networks

### PR DESCRIPTION
**Pull request by aepifanov**
**- What I did**

This PR introduces IP utilization tracking for Docker networks by extending the network inspection API to include IPAM state. This enhancement allows users to monitor how many IP addresses are allocated per subnet and IP range pool within a network.
It covers only IPv4.

**- How I did it**

[Design IP-utilization](https://docs.google.com/document/d/1CmmCe1T8ze8tAOyl3MKsJj_1hVlzpSGK7D4kclS9PW0/edit?tab=t.0)

1. Added IP allocation counters to track usage within:
    - The entire subnet
    - The defined IP range pool
2. Introduced new API types to represent network `State`, including nested `IPAM` state information
3. Implemented support and test coverage for the following network drivers:
    - `bridge`
    - `macvlan`
    - `ipvlan`
    - `overlay`

Example output from `docker network inspect`:
```bash
> docker network inspect my
[
    {
        "Name": "my",
....
        "IPAM": {
            "Driver": "default",
            "Options": {},
            "Config": [
                {
                    "Subnet": "192.168.10.0/24",
                    "IPRange": "192.168.10.128/31",
                    "Gateway": "192.168.10.1",
                    "AuxiliaryAddresses": {
                        "my-host": "192.168.10.128"
                    }
                }
            ]
        },
        "State": {
            "IPAM": {
                "192.168.10.0/24": {
                    "AllocatedIPsInSubnet": 4,
                    "AvailableIPsInIPRangePool": 1
                }
            }
        },
....
    }
]
```

- **AllocatedIPsInSubnet**: Number of addresses allocated in the subnet. If the value exceeds `uint64`, it is capped at the maximum `uint64` value.

- **AvailableIPsInIPRangePool**: Number of available addresses for allocating in the IP range pool.

Depends on:

[IP utilization in Swarm](https://github.com/moby/swarmkit/pull/3209)

**- How to verify it**

Run `docker network inspect <network>` and verify the new `State.IPAM` section reflects real-time allocation counts for each subnet configuration.

**- Human readable description for the release notes**

```markdown changelog
Implements IP utilization tracking for Docker networks.
Adds IP allocation counters to network inspect responses.
```

**- A picture of a cute animal (not mandatory but encouraged)**

image

### CODE REVIEW

**Review by corhere — COMMENTED** [review]
How come you need both the `PoolData.allocatedIPsInPool` counters and `(*addrset.AddrSet).Selected()`? Aren't all reserved IPs added to an AddrSet? I would have expected `.Selected()` to be sufficient.

**Comment by aepifanov** [comment]
@corhere 
> How come you need both the `PoolData.allocatedIPsInPool` counters and `(*addrset.AddrSet).Selected()`? Aren't all reserved IPs added to an AddrSet? I would have expected `.Selected()` to be sufficient.

`allocatedIPsInPool` counts the IP addresses allocated specifically from the IP range pool
`allocatedIPsInSubnet` tracks allocations from the entire subnet

**Comment by corhere** [comment]
> `allocatedIPsInPool` counts the IP addresses allocated specifically from the IP range pool
> `allocatedIPsInSubnet` tracks allocations from the entire subnet

Could you please elaborate further? I still do not understand why the information from the AddrSet is incomplete.

**Comment by aepifanov** [comment]
> > `allocatedIPsInPool` counts the IP addresses allocated specifically from the IP range pool
> > `allocatedIPsInSubnet` tracks allocations from the entire subnet
> 
> Could you please elaborate further? I still do not understand why the information from the AddrSet is incomplete.

`allocatedIPsInSubnet` matches 1:1 with `.Selected()` from the `AddrSet`, as it counts all IPs allocated from the entire subnet.
`allocatedIPsInPool` is calculated based on matches against the configured `ip-range`, which is defined at the `PoolData` level and stored as a `children` map:
https://github.com/moby/moby/blob/addc37324a3aa32c3b74c911ddf92e63b8553aa4/daemon/libnetwork/ipams/defaultipam/structures.go#L18-L25

**Comment by corhere** [comment]
Okay, I think I understand. I am not entirely thrilled with there being two sources of truth for the IP address allocations, but you may have had a good reason to take this approach. Which alternative solutions did you consider and why did the approach of maintaining a running counter win out?

**Comment by aepifanov** [comment]
@corhere I've done the following changes from the last your review:
1. removed ipv6 support
2. reversed the counter for ip-range pool from used to available
3. extened the integratio tests with using aux-addresses
4. changed the State.IPAM format to:
```
        "State": {
            "IPAM": {
                "192.168.10.0/24": {
                    "AllocatedIPsInSubnet": 5,
                    "AvailableIPsInIPRangePool": 1
                }
        },
```

File: daemon/libnetwork/internal/addrset/addrset.go [hunk 1]
```diff
@@ -174,6 +175,35 @@ func (as *AddrSet) getBitmap(addr netip.Addr) (*bitmap.Bitmap, netip.Prefix, err
...
+	// If the subnet is bigger than Bitmap's capacity, calculate the number of bitmaps we have.
+	if bits > maxBitsPerBitmap {
+		bms = 1 << (bits - maxBitsPerBitmap)
+	}
+
+	// Calculate the number of selected addresses in each bitmap.
+	for _, bm := range as.bitmaps {
+		if bms > 0 {
+			bms--
+		}
```
Comment (corhere) [inline daemon/libnetwork/internal/addrset/addrset.go]: What's the purpose of the `bits` and `bms` variables? Neither one is used in the summation.

File: daemon/libnetwork/internal/addrset/addrset.go [hunk 2]
```diff
@@ -174,6 +175,35 @@ func (as *AddrSet) getBitmap(addr netip.Addr) (*bitmap.Bitmap, netip.Prefix, err
...
 }
 
+// Selected returns the number of selected addresses in the set.
+// If the amount of selected addresses is greater than the maximum uint64, it returns the maximum uint64.
+func (as *AddrSet) Selected() (selected uint64) {
+	bits := as.pool.Addr().BitLen() - as.pool.Bits()
+	// If there are no bitmaps, all addresses are unselected.
+	if len(as.bitmaps) == 0 {
+		return 0
+	}
```
Comment (corhere) [inline daemon/libnetwork/internal/addrset/addrset.go]: Nit: premature optimization. The loop body would execute zero times, leading to the correct answer of zero addresses selected without any wasted cycles.

File: daemon/libnetwork/ipams/defaultipam/allocator.go [hunk 3]
```diff
@@ -209,25 +209,32 @@ func (a *Allocator) getAddrSpace(as string, v6 bool) (*addrSpace, error) {
 	return nil, types.InvalidParameterErrorf("cannot find address space %s", as)
 }
 
-func newPoolData(pool netip.Prefix) *PoolData {
-	h := addrset.New(pool)
+func newPoolData(pool, sub netip.Prefix) *PoolData {
```
Comment (corhere) [inline daemon/libnetwork/ipams/defaultipam/allocator.go]: What is the rationale behind changing the signature of `newPoolData`?

File: daemon/libnetwork/ipams/defaultipam/allocator.go [hunk 4]
```diff
@@ -209,25 +209,32 @@ func (a *Allocator) getAddrSpace(as string, v6 bool) (*addrSpace, error) {
...
 
+	}
 	// Reserve the first address in the range for the:
 	// - IPv4 network address
 	//   - Except in a /31 point-to-point link, https://datatracker.ietf.org/doc/html/rfc3021
 	// - IPv6 Subnet-Router anycast address, https://datatracker.ietf.org/doc/html/rfc4291#section-2.6.1
 	bits := pool.Addr().BitLen() - pool.Bits()
 	if !pool.Addr().Is4() || bits > 1 {
-		h.Add(pool.Addr())
+		pd.RequestAddress(pool, netip.Prefix{}, netip.Addr{}, nil)
```
Comment (corhere) [inline daemon/libnetwork/ipams/defaultipam/allocator.go]: How is this equivalent to `pd.addrs.Add(pool.Addr())` when you aren't passing `pool.Addr()` as a parameter?

File: daemon/libnetwork/ipams/defaultipam/structures.go [hunk 5]
```diff
@@ -21,6 +22,9 @@ type PoolData struct {
 	addrs    *addrset.AddrSet
 	children map[netip.Prefix]struct{}
 
+	// The number of addresses allocated in the pool.
+	allocatedIPsInPool uint64
```
Comment (corhere) [inline daemon/libnetwork/ipams/defaultipam/structures.go]: From our earlier discussions I understand that this is not actually the count of addresses allocated in the `nw` as a whole, since that would be duplicating `(PoolData).addrs.Selected()`. What is counting, then? I don't see how a single integer could be used to report how many addresses are available for dynamic allocation in a specific `sub` without interference from other `sub`s under the same `nw` shared with other container networks.

File: daemon/libnetwork/ipams/defaultipam/allocator.go [hunk 6]
```diff
@@ -328,3 +328,25 @@ func getAddress(base netip.Prefix, addrSet *addrset.AddrSet, prefAddress netip.A
...
+// GetAllocatedIPs returns the number of addresses allocated in both the subnet and the ip-range pool.
+//   - allocatedIPsInSubnet: The number of addresses allocated in the subnet.
+//     This value is limited by the uint64 capacity; if the actual count exceeds this limit,
+//     the maximum uint64 value is returned.
+//   - allocatedIPsInPool: The number of addresses allocated in the ip-range pool.
+//     Also limited by uint64 capacity; if it reaches the maximum value, it is treated as frozen
+//     and should no longer be modified.
+func (a *Allocator) GetAllocatedIPs(poolID string) (allocatedIPsInSubnet, allocatedIPsInPool uint64, err error) {
+	log.G(context.TODO()).Debugf("GetAllocatedIPs(%s)", poolID)
+	k, err := PoolIDFromString(poolID)
```
Comment (corhere) [inline daemon/libnetwork/ipams/defaultipam/allocator.go]: I don't see `k.ChildSubnet` consumed anywhere. How can the count of addresses allocated in that range be looked up without knowing which range to count?

File: daemon/libnetwork/network.go [hunk 7]
```diff
@@ -2193,3 +2194,59 @@ func (n *Network) deleteLoadBalancerSandbox() error {
...
+			continue
+		}
+
+		// Calculate the capacity of the pool, which is currently just a CIDR.
+		// Only IPv4 is calculated, as the IPAM state only supports IPv4.
+		capacity := func(cidr netip.Prefix) (uint64, error) {
+			if !cidr.Addr().Is4() {
+				return 0, fmt.Errorf("IPAM state only supports IPv4 CIDRs, got %s", cidr)
+			}
+			return uint64(1) << (32 - cidr.Bits()), nil
```
Comment (corhere) [inline daemon/libnetwork/network.go]: We can do better than hardcode a constant:
```suggestion
			return uint64(1) << (cidr.Addr().BitLen() - cidr.Bits()), nil
```

File: daemon/libnetwork/ipams/defaultipam/address_space_test.go [hunk 8]
```diff
@@ -498,3 +498,119 @@ func TestPoolAllocateAndRelease(t *testing.T) {
...
+			},
+		},
+	}
+
+	for _, tc := range testcases {
+		t.Run(tc.name, func(t *testing.T) {
+			as, err := newAddrSpace(tc.predefined)
+			assert.NilError(t, err)
+
+			err = as.allocateSubnet(tc.subnet, tc.ipRange)
```
Comment (corhere) [inline daemon/libnetwork/ipams/defaultipam/address_space_test.go:554]: Where are you testing the cases with multiple ip-range allocations under a single subnet? Be sure to cover both the disjoint ranges (e.g. `--subnet 192.168.0.0/24 --ip-range 192.168.0.0/25`, `--subnet 192.168.0.0/24 --ip-range 192.168.0.128/25`) [and overlapping ranges](https://github.com/moby/moby/issues/46756) (e.g. `--subnet 192.168.0.0/24 --ip-range 192.168.0.0/25`, `--subnet 192.168.0.0/24 --ip-range 192.168.0.64/26`) cases

**Review by corhere — CHANGES_REQUESTED** [review]
It does not look like the tests cover all the corner cases.

File: daemon/libnetwork/internal/addrset/addrset.go (hunk 1 above)
Comment (aepifanov) [reply daemon/libnetwork/internal/addrset/addrset.go]: Outdated code. Removed.

File: daemon/libnetwork/ipams/defaultipam/allocator.go (hunk 3 above)
Comment (aepifanov) [reply daemon/libnetwork/ipams/defaultipam/allocator.go]: The idea was to handle `usedInRangePool` counter during the PoolData initialization steps, which required the `sub` info as well. 
But it's outdated code for now.

File: daemon/libnetwork/ipams/defaultipam/allocator.go (hunk 4 above)
Comment (aepifanov) [reply daemon/libnetwork/ipams/defaultipam/allocator.go]: After re-implementation of `usedInIPRangePool` calculation it's returned back.

File: daemon/libnetwork/ipams/defaultipam/structures.go (hunk 5 above)
Comment (aepifanov) [reply daemon/libnetwork/ipams/defaultipam/structures.go]: Nice catch!
I've re-implemented `usedInIPRangePool` handling method to the direct calculation of the selected addresses in a bitmap for a specified range, which allows to get the actual IPAM state for any range.

File: daemon/libnetwork/network.go (hunk 7 above)
Comment (aepifanov) [reply daemon/libnetwork/network.go]: Done

File: daemon/libnetwork/ipams/defaultipam/allocator.go (hunk 6 above)
Comment (aepifanov) [reply daemon/libnetwork/ipams/defaultipam/allocator.go]: In the current implementation the `range` is taken from the `PoolID` which contains it.

File: daemon/libnetwork/ipams/defaultipam/address_space_test.go (hunk 8 above)
Comment (aepifanov) [reply daemon/libnetwork/ipams/defaultipam/address_space_test.go:554]: Added test-cases.

File: daemon/cluster/convert/network.go [hunk 9]
```diff
@@ -155,6 +158,15 @@ func BasicNetworkFromGRPC(n swarmapi.Network) network.Inspect {
...
 	}
+	netState := &network.NetworkState{}
+	if n.State != nil {
+		if err := json.Unmarshal(n.State.Value, netState); err != nil {
+			log.G(context.TODO()).WithError(err).Warnf("Failed to unmarshal network state for network %s", n.ID)
+		}
+	}
+	if netState.IPAM == nil {
+		netState = nil
+	}
```
Comment (corhere) [inline daemon/cluster/convert/network.go]: The daemon instance converting the `swarmapi.Network` to a `network.Inspect` might not be the same process, let alone the same version, as the one that produced the `swarmapi.Network` value. Could you please future-proof the conversion to gracefully handle network state values of unexpected types? Having the consumer assume a particular schema of the `Any` value hampers our flexibility to change how the network state is encoded.

File: daemon/libnetwork/bitmap/sequence.go [hunk 10]
```diff
@@ -176,6 +180,72 @@ func (s *sequence) fromByteArray(data []byte) error {
...
+	for remaining > 0 {
+		if cur.count <= remaining {
+			selected += uint64(bits.OnesCount32(cur.block)) * cur.count
+			remaining -= cur.count
+			cur = cur.next
+		} else {
+			selected += uint64(bits.OnesCount32(cur.block)) * remaining
+			remaining = 0
+		}
+	}
```
Comment (corhere) [inline daemon/libnetwork/bitmap/sequence.go]: Recent versions of Go now have convenient [`max`](https://pkg.go.dev/builtin@go1.24.5#min) and [`min`](https://pkg.go.dev/builtin@go1.24.5#min) builtins.

```suggestion
	for remaining > 0 {
		n := min(cur.count, remaining)
		selected += uint64(bits.OnesCount32(cur.block)) * n
		remaining -= n
		cur = cur.next
	}
```

File: daemon/libnetwork/bitmap/sequence.go [hunk 11]
```diff
@@ -176,6 +180,72 @@ func (s *sequence) fromByteArray(data []byte) error {
...
+
+	// Last block processing
+	selected += uint64(bits.OnesCount32(eCurrent.block & loBlockMask(endBit)))
+
+	return selected, nil
+}
+
+// CalculateSelected calculates the number of selected bits in the range [start, end].
+func (h *Bitmap) CalculateSelected(start, end uint64) (uint64, error) {
```
Comment (corhere) [inline daemon/libnetwork/bitmap/sequence.go]: Bikeshedding: name functions based on their outputs and side effects, not on irrelevant implementation details. It makes no difference to the caller whether the number of set bits is calculated on the fly, cached, memoized, indexed, etc., so long as it gets the correct answer.
```suggestion
// OnesCount calculates the number of selected bits in the range [start, end].
func (h *Bitmap) OnesCount(start, end uint64) (uint64, error) {
```

File: daemon/libnetwork/bitmap/sequence.go [hunk 12]
```diff
@@ -590,3 +660,19 @@ func ordinalToPos(ordinal uint64) (uint64, uint64) {
...
+// hiBlockMask generates a mask for the high bits of a block based on the bit position including this bit position.
+func hiBlockMask(bitPos uint64) uint32 {
+	if bitPos >= uint64(blockLen) {
+		return 0
+	}
+	if bitPos == 0 {
+		return blockFull
+	}
+	return (blockFirstBit >> (bitPos - 1)) - 1
+}
```
Comment (corhere) [inline daemon/libnetwork/bitmap/sequence.go]:
```suggestion
func hiBlockMask(bitPos uint64) uint32 {
	return blockFull >> bitPos
}
```

File: daemon/libnetwork/bitmap/sequence_test.go [hunk 13]
```diff
@@ -1231,3 +1234,77 @@ func TestMarshalJSON(t *testing.T) {
...
+	assert.Check(t, is.Equal(uint32(0xe0000000), loBlockMask(2)))
+	assert.Check(t, is.Equal(uint32(0xffff0000), loBlockMask(15)))
+	assert.Check(t, is.Equal(uint32(0xffff8000), loBlockMask(16)))
+	assert.Check(t, is.Equal(uint32(0xfffffffe), loBlockMask(30)))
+	assert.Check(t, is.Equal(uint32(0xffffffff), loBlockMask(31)))
+	assert.Check(t, is.Equal(uint32(0xffffffff), loBlockMask(32)))
+	assert.Check(t, is.Equal(uint32(0xffffffff), loBlockMask(33)))
+}
+
+func TestCalculateSelected(t *testing.T) {
```
Comment (corhere) [inline daemon/libnetwork/bitmap/sequence_test.go]: I would trust the implementation a lot more if CalculateSelected was tested by setting up the preconditions through the public interface of bitmap. By directly constructing the sequences it's possible that an incorrect assumption in the code could be carried through to the test cases such that the tests don't catch the bug. This test as written would not catch if you mixed up the bit-endianness of the blocks, for instance.

File: daemon/libnetwork/network.go [hunk 14]
```diff
@@ -2190,3 +2191,46 @@ func (n *Network) deleteLoadBalancerSandbox() error {
...
+
+	ipam4Infos, _ := n.IpamInfo()
+	// Only IPv4 is calculated, as the IPAM state only supports IPv4.
+	if len(ipam4Infos) == 0 {
+		return nil
+	}
+
+	ipamState := map[string]network.IPAMState{}
+	for _, ii := range ipam4Infos {
+		cidr, is, err := defaultipam.GetIPAMStateForPoolID(ii.PoolID, ipam)
```
Comment (corhere) [inline daemon/libnetwork/network.go]: Why is generic libnetwork code for any network calling a function from the default IPAM driver? What happens if the IPAM driver for the network is not `defaultipam`?

File: go.mod [hunk 15]
```diff
@@ -270,3 +270,5 @@ exclude (
 replace github.com/moby/moby/api => ./api
 
 replace github.com/moby/moby/client => ./client
+
+replace github.com/moby/swarmkit/v2 v2.0.0 => github.com/aepifanov/swarmkit/v2 v2.0.0-20250805114649-a1269cd7a2a8
```
Comment (corhere) [inline go.mod]: We can't merge this PR with the Swarmkit dependency replaced with your fork. Please convert this PR to draft.

File: daemon/cluster/convert/network.go [hunk 16]
```diff
@@ -155,6 +158,35 @@ func BasicNetworkFromGRPC(n swarmapi.Network) network.Inspect {
...
+					}
+				}
+			default:
+				log.G(context.TODO()).WithError(err).Warnf("network state for network %s has unsupported version  %d", n.ID, env.Version)
+				if env.Version > 1 && env.Data != nil {
+					log.G(context.TODO()).WithError(err).Warnf("Try to unmarshal it anyway since the version is higher %d than supported %d and might be backward compatible for network %s", env.Version, 1, n.ID)
+					if err := json.Unmarshal(env.Data, netState); err != nil {
+						log.G(context.TODO()).WithError(err).Warnf("Failed to unmarshal network state for network %s", n.ID)
+					}
+				}
```
Comment (corhere) [inline daemon/cluster/convert/network.go]: What's the point of versioning the type if you're just going to proceed anyway? Why would we bump the version if it's backwards-compatible with v1? And if the type is not backwards-compatible, it's a different data type by definition, so changing the type name would be the more appropriate action than bumping the version.

File: daemon/cluster/convert/network.go [hunk 17]
```diff
@@ -155,6 +158,35 @@ func BasicNetworkFromGRPC(n swarmapi.Network) network.Inspect {
 			})
 		}
 	}
+	netState := &network.NetworkState{}
+	if n.State != nil {
+		env := network.StringifyEnvelope{}
+		if err := json.Unmarshal(n.State.Value, &env); err != nil {
```
Comment (corhere) [inline daemon/cluster/convert/network.go]: How does the StringifyEnvelope help to enable us to switch to another codec, e.g. protobuf or msgpack, for the state value in the future? It looks like this will lock us into JSON forever.

File: daemon/cluster/convert/network.go [hunk 18]
```diff
@@ -155,6 +158,35 @@ func BasicNetworkFromGRPC(n swarmapi.Network) network.Inspect {
...
+			log.G(context.TODO()).WithError(err).Warnf("Failed to unmarshal network state for network %s", n.ID)
+		} else if env.Type != network.NetworkStateType {
+			log.G(context.TODO()).WithError(err).Warnf("network state for network %s has unexpeted message type %s", n.ID, env.Type)
```
Comment (corhere) [inline daemon/cluster/convert/network.go]: What can we infer about how this manager node relates to the leader when this code path is taken? Is there something actionable we could include in the log message?

File: daemon/libnetwork/cnmallocator/networkallocator.go [hunk 19]
```diff
@@ -980,3 +984,51 @@ func setIPAMSerialAlloc(opts map[string]string) map[string]string {
...
+		}
+		stringifyEnvelopeState, err := json.Marshal(&envelopedState)
+		if err != nil {
+			return errors.Wrapf(err, "failed to marshal network state for network %s", net.ID)
+		}
+		net.State = &gogotypes.Any{
+			TypeUrl: "types.docker.com/NetworkState",
+			Value:   stringifyEnvelopeState,
+		}
+	}
```
Comment (corhere) [inline daemon/libnetwork/cnmallocator/networkallocator.go]: Please co-locate the marshalling code in the same package as the corresponding unmarshalling code. I recommend putting it in daemon/cluster/convert/network.go, the same file as the unmarshaller and same package as all the other code which converts between Engine and Swarm API types.

File: api/types/network/network.go [hunk 20]
```diff
@@ -80,6 +99,7 @@ type Inspect struct {
 	EnableIPv4 bool                        // EnableIPv4 represents whether IPv4 is enabled
 	EnableIPv6 bool                        // EnableIPv6 represents whether IPv6 is enabled
 	IPAM       IPAM                        // IPAM is the network's IP Address Management
+	State      *NetworkState               `json:",omitempty"` // State represents the state of the network
```
Comment (corhere) [inline api/types/network/network.go:88]: New fields added to API responses need to be gated on the API version so that people developing clients are forced to set the API version correctly and therefore have their clients fail fast when connecting to incompatible engines. Please modify the API endpoints to clear this field on API versions less than the latest (v1.52)

Comment (corhere) [inline api/types/network/network.go:88]: Please update the Swagger spec and API changelog with this addition

**Comment by corhere** [comment]
- Superseded by #50917 and #50946
